chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -2127,14 +2133,92 @@ impl PeerManagerActor {
+            // we should try to sign data and broadcast it. However, this is
+            // also a trigger for a full sync, so a dedicated broadcast is
+            // not required.
+            //
+            // TODO(gprusak): For dynamic self-IP-discovery, add a STUN daemon which
+            // will add new AccountData and trigger an incremental broadcast.
+            if let Some(vc) = &state.config.validator {
+                let my_account_id = vc.signer.validator_id();
+                let my_public_key = vc.signer.public_key();
+                // TODO(gprusak): STUN servers should be queried periocally by a daemon
+                // so that the my_peers list is always resolved.
+                // Note that currently we will broadcast an empty list.
+                // It won't help us to connect the the validator BUT it
+                // will indicate that a validator is misconfigured, which
+                // is could be useful for debugging. Consider keeping this
+                // behavior for situations when the IPs are not known.
+                let my_peers = match &vc.endpoints {
+                    config::ValidatorEndpoints::TrustedStunServers(_) => vec![],
+                    config::ValidatorEndpoints::PublicAddrs(addrs) => {
+                        addrs.iter().cloned().map(|addr| PeerAddr { addr, peer_id: None }).collect()
+                    }
+                };
+                let my_data = info.tier1_accounts.iter().filter_map(|((epoch_id,account_id),key)| {
+                    if account_id != my_account_id{
+                        return None;
+                    }
+                    if key != &my_public_key {
+                        warn!(target: "network", "node's account_id found in TIER1 accounts, but the public keys do not match");
+                        return None;
+                    }
+                    Some(AccountData {
+                        epoch_id: epoch_id.clone(),
+                        account_id: my_account_id.clone(),
+                        timestamp: now,
+                        peers: my_peers.clone(),
+                    }.sign(vc.signer.as_ref()).unwrap())
+                }).collect();
+                // Insert node's own AccountData should never fail.
+                // We ignore the new data, because we trigger inserting
+                if let (_, Some(err)) = state.accounts_data.clone().insert(my_data).await {
+                    panic!("inserting node's own AccountData to self.state.accounts_data: {err}");
+                }
+            }
+            // The set of tier1 accounts has changed.
+            // We might miss some data, so we start a full sync with the connected peers.
+            // TODO(gprusak): add a daemon which does a periodic full sync in case some messages
+            // are lost (at a frequency which makes the additional network load negligible).
+            state
```

Comment:
I am a bit lost here: why are we sending `SyncAccountsData` if we didn't enter the `if let Some(vc) = &state.config.validator` if before? I assume for the `equesting_full_sync: true,` bit? Could you add a comment on `SyncAccountsData` explaining how this works? My understanding is that `SyncAccountsData` is both a data itself, as well as a request to get this same data back. This ... is a bit mind-bending, so perhaps some comment on `SyncAccountsData` would help? 

Alternatively, perhaps there's a better name for `requesting_full_sync`? I was under the impression that it is incermental/full rather than one-way/two-way

Reply:
It supports both modes, because both are needed: if the set of tier1 accounts changes we need to request the data for new accounts somehow, because we might have ignored them when it was broadcasted previously. This request doesn't technically have to contain any accounts_data, I've just added it for consistency.

It is true that I've merged 3 message types into 1 (to save on number of messages sent): FullSyncRequest, FullSyncResponse, IncrementalSync. Do you think I should split them back?

Reply:
>Do you think I should split them back?

No, I actually find the single message design quite elegant! But, as with many intricately-elegant designs, it could benefit from a strategically placed comment :)

Reply:
there is a comment in the network.proto file. I will add a reference to it in network_protocol/mod.rs